cloud-controller-manager/do/kubernetes_lb_validator.go

```diff
@@ -0,0 +1,138 @@
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package do
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/digitalocean/godo"
+	"github.com/go-logr/logr"
+	v1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+// LBService represents the v1.service lb object data
+type LBService struct {
+	APIVersion string
+	Kind       string
+	Metadata   Metadata
+	Spec       map[string]interface{}
+	Selector   map[string]interface{}
+	Type       string
+}
+
+// Metadata represents the metadata field in the lb service object
+type Metadata struct {
+	Annotations interface{} `json:"annotation data,omitempty"`
+	Name        string
+	Namespace   string
+}
+
+// DOKSLBServiceValidator validates service type LB
+type KubernetesLBServiceValidator struct {
+	decoder   *admission.Decoder
+	Log       logr.Logger
+	GClient   *godo.Client
+	ClusterID string
+	Region    string
+}
+
+// DOKSLBServiceValidator ...
+func (v *KubernetesLBServiceValidator) Handle(ctx context.Context, req admission.Request) admission.Response {
+	svc := &v1.Service{}
+
+	v.Log.V(6).Info("decoding received request")
+	err := v.decoder.Decode(req, svc)
+	if err != nil {
+		v.Log.Error(err, "failed to decode request")
+		return admission.Errored(http.StatusBadRequest, fmt.Errorf("failed to decode request: %v", err))
+	}
+
+	v.Log.V(6).Info("checking received request")
+	if svc.Spec.Type != v1.ServiceTypeLoadBalancer || svc.DeletionTimestamp != nil {
+		return admission.Allowed("the service object is either not a load balancer or being deleted")
+	}
+
+	tags := []string{
+		"k8s",
+		"k8s:worker",
+		fmt.Sprintf("k8s: %v", v.ClusterID),
+	}
+
+	// TODO: these forwarding rules are a placeholder. Further development is required to extract the values from the
+	//svc object
+	forwardingRules := []godo.ForwardingRule{
+		{
+			EntryProtocol:  "http",
+			EntryPort:      80,
+			TargetProtocol: "http",
+			TargetPort:     80,
+			CertificateID:  "",
+			TlsPassthrough: false,
+		}}
+
+	lbRequest := v.buildRequest(svc.Name, v.Region, tags, forwardingRules)
+
+	err = v.decoder.DecodeRaw(req.OldObject, svc)
+	if err != nil {
+		err = v.validateCreate(ctx, lbRequest)
+		if err != nil {
+			v.Log.Error(err, "failed to validate lb creation")
+			return admission.Denied("failed to validate lb creation")
+		}
+		return admission.Allowed("valid lb create request")
+	}
+
+	currentLBID := svc.Annotations["kubernetes.digitalocean.com/load-balancer-id"]
+	if currentLBID == "" {
+		v.Log.Error(fmt.Errorf("missing loadbalancer id in annotation"), "missing kubernetes.digitalocean.com/load-balancer-id annotation")
```

Comment:
If the annotation is missing, then it means we need to do a create request as well.